pkg/controller/apiextensions/composed/reconciler.go

```diff
@@ -0,0 +1,126 @@
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	runtimev1alpha1 "github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+
+	"github.com/crossplane/crossplane/apis/apiextensions/v1alpha1"
+	"github.com/crossplane/crossplane/apis/apiextensions/v1alpha1/instance"
+)
+
+const (
+	shortWait = 30 * time.Second
+	longWait  = 1 * time.Minute
+	timeout   = 2 * time.Minute
+)
+
+type ConnectionSecretFilterer interface {
+	GetConnectionSecretKeys() []string
+}
+
+// NewInstanceReconciler returns a new *instanceReconciler.
+func NewInstanceReconciler(name string, mgr manager.Manager, gvk schema.GroupVersionKind, log logging.Logger, filterer ConnectionSecretFilterer) reconcile.Reconciler {
+	ni := func() instance.CompositedInstance {
+		cr := &instance.InfraInstance{}
+		cr.SetGroupVersionKind(gvk)
+		return cr
+	}
+	kube := NewClientForUnregistered(mgr.GetClient())
+	return &instanceReconciler{
+		client:      kube,
+		newInstance: ni,
+		Resolver:    NewSelectorResolver(kube),
+		composited:  NewAPICompositedResourcesReconciler(kube),
+		connection:  NewAPIFilteredSecretPublisher(kube, filterer.GetConnectionSecretKeys()),
+		log:         log,
+		record:      event.NewAPIRecorder(mgr.GetEventRecorderFor(name)),
+	}
+}
+
+// Resolver lists the operations that are done on the spec of instance.
+type Resolver interface {
+	ResolveSelector(ctx context.Context, cr instance.CompositedInstance) error
+}
+
+type CompositedResourcesReconciler interface {
+	Reconcile(ctx context.Context, cr instance.CompositedInstance, comp v1alpha1.Composition) (managed.ConnectionDetails, error)
+}
+
+// instanceReconciler reconciles the generic CRD that is generated via InfrastructureDefinition.
+type instanceReconciler struct {
+	client      client.Client
+	newInstance func() instance.CompositedInstance
+	composited  CompositedResourcesReconciler
+	connection  managed.ConnectionPublisher
+
+	Resolver
+	resource.Finalizer
+
+	log    logging.Logger
+	record event.Recorder
+}
+
+// Reconcile reconciles given custom resource.
+func (r *instanceReconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) { // nolint:gocyclo
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	cr := r.newInstance()
+	if err := r.client.Get(ctx, req.NamespacedName, cr); err != nil {
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), "cannot get instance")
+	}
+
+	if err := r.ResolveSelector(ctx, cr); err != nil {
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(err, "cannot resolve composition selector")
+	}
```

Comment:
I couldn't see much upside of having a separate controller, which brings some overhead. Do you see how it'd benefit?

Reply:
Actually, I'm thinking of using your recent selector resolver. It's pretty much the same; choose resources with given GVK using label selector to get its name. What do you think?

Reply:
> I'm thinking of using your recent selector resolver.

Hrm, it's an unexpected usage, but it could work.

> Do you see how it'd benefit?

It's a pretty small benefit. I'd say if there was any case it would be that it would work more like our other existing controllers, and that it would give us space to build more complicated selection logic (e.g. weighted compositions to model price or performance) without adding to the complexity of this controller.

I suspect I'd like the selection to be a separate controller eventually, but it's fine to worry about that in a future PR.

Reply:
I would prefer to separate it when that need comes up. We might even switch to mutating [webhook](https://github.com/crossplane/crossplane/issues/1152) for that stuff in the future.